Trim leading and trailing periods of missing data

wfvining opened the pull request:
Many data-sets come with leading and trailing periods that have sporadic or completely missing data. This adds two functions for identifying and removing these periods:
* `quality.gaps.valid_between()` gives the first and last valid dates in the series.
* `quality.gaps.trim()` returns a boolean mask that is True for the valid period between the dates given by `valid_between`.

These functions both operate on days in the data set, marking all data for each day as valid/invalid based on the number of hours of data that exists on each day. Any data point with a value other than `NaN` is considered "valid" data.

cwhanse commented on pvanalytics/quality/gaps.py:
```diff
@@ -135,3 +136,106 @@ def interpolation_diff(x, window=3, rtol=1e-5, atol=1e-8):
+    consecutive days with data covering at least `minimum_hours`
+    preceeding it.
+
+    Any data point with a value other than `NaN` is considered valid
```
Since we are only removing `NaN`, "valid" is a bit strong.

I would use "valid" for data that also passes physical limit and consistency checks, for example. I also wouldn't use "complete" or "serially complete" since I would expect the output to not have _any_ missing values. Maybe "sufficient"? 

If `series` is 30 days, and there are lots of missing data on day 1, day 15, and day 30, what is returned?  Day 2, Day 14?  Or something else?

wfvining replied:
I think the important thing to describe is what it means for a value to be missing rather than what it means for it to be 'valid.' You are right, "valid" is not the best word to use here.

> If series is 30 days, and there are lots of missing data on day 1, day 15, and day 30, what is returned? Day 2, Day 14? Or something else?

With the default parameters the returned mask will select day 2-29. If there is missing data in the middle of a series nothing happens to it.

wfvining commented:
Per discussion with Matt and Cliff should expand this to include a function that gives a completeness score to each day (hours per day with data) and a function that gives a boolean mask for days with completeness > some threshold. This mask can/should be used in `trim()`.

wfvining commented:
Added daily completeness filters and refactored to use these functions in `valid_between`. Renamed `valid_between` to `start_stop_dates` and rewrote documentation to describe the function more accurately.

Tasks
- [ ] add documentation for new functions to API docs
- [ ] fix existing docs (did not change name of `valid_between` there.)

cwhanse commented on pvanalytics/quality/gaps.py:
```diff
@@ -135,3 +136,206 @@ def interpolation_diff(x, window=3, rtol=1e-5, atol=1e-8):
+
+def start_stop_dates(series, days=10, minimum_completeness=0.333333,
+                     freq=None):
+    """Get the start and end of data excluding leading and trailing gaps.
```
Perhaps this function could be made more general by accepting a time series of Boolean, and returning start and stop where the minimum daily count of True is less than the threshold.  That way the criteria for start and stop could change, e.g., fraction of times with irradiance like clear sky.

cwhanse replied:
If we want to keep this function as is, for PVFleets, we could make a `trim_pvfleets` function that implements their workflow.

wfvining replied:
I like the idea of making this more general is good. If you think passing a boolean series in would achieve that, then I will go ahead and make that change. `trim_pvfleets` (or whatever we want to call it) can easily be implemented using the more general function.

wfvining commented on pvanalytics/quality/gaps.py:
```diff
@@ -135,3 +136,206 @@ def interpolation_diff(x, window=3, rtol=1e-5, atol=1e-8):
+    freq_seconds = _freq_to_seconds(freq)
+    seconds_per_sample = freq_seconds or inferred_seconds
+    if freq and inferred_seconds < freq_seconds:
+        raise ValueError("freq must be less than or equal to the"
```
The rationale behind passing `freq` is that the original time series may have been heavily reduced by previous quality control tests, resulting in an inferred frequency that is much longer than the actual frequency of the data. In that case we would attribute too much time to each data point and calculate a higher completeness than we actually have.

cwhanse replied:
OK, let's roll with the separate `freq` argument and see how this gets used.

wfvining commented on pvanalytics/quality/gaps.py:
```diff
@@ -135,3 +136,206 @@ def interpolation_diff(x, window=3, rtol=1e-5, atol=1e-8):
+    """
+    inferred_seconds = _freq_to_seconds(pd.infer_freq(series.index))
+    freq_seconds = _freq_to_seconds(freq)
+    seconds_per_sample = freq_seconds or inferred_seconds
```
The `or` operator short-circuits, returning `freq_seconds` in that case. I think it isn't clear because as you read the code you assume `_freq_to_seconds` always returns a value (if its argument is `None` then it returns `None`). I will re-work it for clarity.

cwhanse replied:
Or put some inline comments for future maintainers

wfvining replied:
I ended up rewriting it. I think there is just a little too much going on for the `arg or default` pattern to be clear.

wfvining commented on pvanalytics/quality/gaps.py:
```diff
@@ -135,3 +136,206 @@ def interpolation_diff(x, window=3, rtol=1e-5, atol=1e-8):
+
+    """
+    complete_days = daily_completeness(series, freq) >= minimum_completeness
+    return complete_days.reindex(series.index, method='pad')
```
I will address this with a kwarg to `daily_completeness` (and rename the function to `completeness_score`).

wfvining commented:
Added a generic trim function that operates on a series of booleans.

cwhanse commented on pvanalytics/quality/gaps.py:
```diff
@@ -135,3 +136,209 @@ def interpolation_diff(x, window=3, rtol=1e-5, atol=1e-8):
+def trim_incomplete(series, minimum_completeness=0.333333, days=10, freq=None):
+    """Mask out missing data from the beginning and end of the data.
+
+    Removes data preceeding the start date and following the stop date
```
```suggestion
    False for times preceeding the start date and following the stop date
```

wfvining commented on docs/api.rst:
```diff
@@ -53,6 +53,23 @@ Identify gaps in the data.
+   :toctree: generated/
+
+   quality.gaps.start_stop_dates
+   quality.gaps.trim
```
Add `trim_incomplete` here
